### ISSUE TITLE
Using Enum dynamic values as a Tuple inner trait does not perform expected validation

### ISSUE THREAD

**Issue by kitchoi** [enthought/traits#1385]
Given the following object:
```python
class Bar(HasStrictTraits):

    names = Set(Str())

    name_and_enum = Tuple(Str(), Enum(values="names"))

    enum_and_name = Tuple(Enum(values="names"), Str())

    enum_and_enum = Tuple(Enum(values="names"), Enum(values="names"))
```

I'd expect the `name_and_enum` to raise an error if the tuple's second element is not one of the values in `names`. Likewise, `enum_and_name` should raise an error if the tuple's first element is not one of the values in `names`. 
But it seems not to be the case, if I run these tests:

```python

class TestBar(unittest.TestCase):

    def test_name_and_name_ok(self):
        bar = Bar(names={"a", "b", "c"})
        bar.enum_and_name = ("a", "b")

    def test_name_and_enum_ok(self):
        bar = Bar(names={"a", "b", "c"})
        bar.name_and_enum = ("a", "b")

    def test_name_and_enum_not_ok_second_element(self):
        bar = Bar(names={"a", "b", "c"})

        with self.assertRaises(TraitError):
            bar.name_and_enum = ("a", "x")

    def test_enum_and_name_not_ok_first_element(self):
        bar = Bar(names={"a", "b", "c"})

        with self.assertRaises(TraitError):
            bar.enum_and_name = ("x", "a")

    def test_enum_and_enum_not_ok_first_element(self):
        bar = Bar(names={"a", "b", "c"})

        with self.assertRaises(TraitError):
            bar.enum_and_enum = ("x", "a")

    def test_enum_and_enum_not_ok_second_element(self):
        bar = Bar(names={"a", "b", "c"})

        with self.assertRaises(TraitError):
            bar.enum_and_enum = ("a", "x")
```

All the `assertRaises(TraitError)` fail.

If I changed `Enum(values="names")` to `Enum("a", "b", "c")` then all the tests pass.

**Comment by mdickinson** [comment]
Looking into this bug ...

**Comment by mdickinson** [comment]
This looks like the result of this line: https://github.com/enthought/traits/blob/4014e841169388bcb885aae3ab7893f40ded335e/traits/trait_types.py#L1979

IOW, it's a fairly simple oversight in `Enum` itself: it's not okay to do all your validation in the `set` method, because there are use-cases for doing validation without actually setting, like the one that caused this bug report.

Should be a fairly easy fix. While we're at it, we should double-check to see if we're setting `self.validate = None` for any other trait types.

**Comment by mdickinson** [comment]
`BaseRange` has the same issue.

**Comment by mdickinson** [comment]
There's a secondary issue here, which is that the tuple validation (`validate_trait_tuple_check`) currently silences all exceptions, not just `TraitError`. Changing that does risk code breakage.

**Comment by mdickinson** [comment]
And yet another issue, exhibited by the following unit test. The cause seems to be different from the dynamic-enum-in-tuple case. Investigating.

```python
    def test_dynamic_enum_in_list(self):
        # Another regression test for #1385.
        class HasEnumInList(HasTraits):
            #: Valid digits
            digits = Set(Int)

            #: Sequence of those digits
            digit_sequence = List(Enum(values="digits"))

        model = HasEnumInList(digits={-1, 0, 1})
        model.trinary_int = [-1, 0, 1, 1]
        with self.assertRaises(TraitError):
            model.trinary_int = [-1, 0, 2, 1]
```

**Comment by mdickinson** [comment]
> And yet another issue, exhibited by the following unit test.

Whoops. That was simply a bug in the test - I renamed `trinary_int` in the class but not in the test code that exercised that class.

**Comment by rahulporuri** [comment]
adding this to the 6.2.0 milestone as well.

**Comment by mdickinson** [comment]
Fixed by #1388. I've opened separate related issues arising from the discussion in comments above:

- #1389: Tuple validation  silences exceptions
- #1390: Value getting is problematic for dynamic Traits inside a container
- #1391: Analog of this issue for `BaseRange`

**Comment by mdickinson** [comment]
Turns out that this issue was a duplicate (of #504).
